### PR TITLE
Platform daemon changes for multi asic platform

### PR DESCRIPTION
**Pull request by judyjoseph**
The changes done here are for supporting ledd and xcvr for multi-asic platforms 
(i) In the ledd process, select now looks for events from the APP_DB in different namespaces.
(ii) In the xcvr process, the parent process and the sub processes connect to databases in all namespaces. The table in which namespace to be updated, is decided by the logical port ( we use the logical_port to asic_id mapping which is stored in the sfp_helper utility classes while parsing the port_config.ini files )

Related PR's

https://github.com/Azure/sonic-swss-common/pull/364
https://github.com/Azure/sonic-buildimage/pull/4932
https://github.com/Azure/sonic-platform-common/pull/100

Update with test results in Single ASIC and Multi-ASIC 
-------------------------------------------------------------

**SINGLE ASIC**

```
admin@str-s6000-acs-8:~$ show interface transceiver eeprom
Ethernet0: SFP EEPROM detected
        Connector: No separable connector
        Encoding: Unspecified
        Extended Identifier: Power Class 1(1.5W max)
        Extended RateSelect Compliance: QSFP+ Rate Select Version 1
        Identifier: QSFP+ or later
        Length Cable Assembly(m): 1
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: 
                10/40G Ethernet Compliance Code: 40GBASE-CR4
                Fibre Channel Speed: 1200 Mbytes/Sec
                Fibre Channel link length/Transmitter Technology: Electrical inter-enclosure (EL)
                Fibre Channel transmission media: Twin Axial Pair (TW)
        Vendor Date Code(YYYY-MM-DD Lot): 2016-05-07 
        Vendor Name: Molex Inc.
        Vendor OUI: 00-09-3a
        Vendor PN: 1110401054
        Vendor Rev: 
        Vendor SN: 612833363

Ethernet4: SFP EEPROM detected
.....
admin@str-s6000-acs-8:~$ show interface transceiver presence 
Port         Presence
-----------  ----------
Ethernet0    Present
Ethernet4    Present
Ethernet8    Present
Ethernet12   Present
Ethernet16   Present
Ethernet20   Present
Ethernet24   Present
Ethernet28   Present
Ethernet32   Present
Ethernet36   Present
Ethernet40   Present
Ethernet44   Present
Ethernet48   Present
Ethernet52   Present
Ethernet56   Present
Ethernet60   Present
Ethernet64   Present
Ethernet68   Present
Ethernet72   Present
Ethernet76   Present
Ethernet80   Present
Ethernet84   Present
Ethernet88   Present
Ethernet92   Present
Ethernet96   Present
Ethernet100  Present
Ethernet104  Present
Ethernet108  Present
Ethernet112  Present
Ethernet116  Present
Ethernet120  Present
Ethernet124  Present

admin@str-s6000-acs-8:~$ sudo config interface transceiver lpmode Ethernet84 enable
Enabling low-power mode for port Ethernet84...  OK

admin@str-s6000-acs-8:~$ sudo config interface transceiver lpmode Ethernet0 disable
Disabling low-power mode for port Ethernet0...  OK

admin@str-s6000-acs-8:~$ show interface transceiver lpmode
Port         Low-power Mode
-----------  ----------------
Ethernet0    Off
Ethernet4    Off
Ethernet8    Off
Ethernet12   Off
Ethernet16   Off
Ethernet20   Off
Ethernet24   Off
Ethernet28   Off
Ethernet32   Off
Ethernet36   Off
Ethernet40   Off
Ethernet44   Off
Ethernet48   Off
Ethernet52   Off
Ethernet56   Off
Ethernet60   Off
Ethernet64   Off
Ethernet68   Off
Ethernet72   Off
Ethernet76   Off
Ethernet80   Off
Ethernet84   On
Ethernet88   Off
Ethernet92   Off
Ethernet96   Off
Ethernet100  Off
Ethernet104  Off
Ethernet108  Off
Ethernet112  Off
Ethernet116  Off
Ethernet120  Off
Ethernet124  Off
```

**Multi-ASIC**

```
root@str-n3164-acs-1:/# ps -ef 
UID        PID  PPID  C STIME TTY          TIME CMD
root         1     0  2 22:03 pts/0    00:00:01 /usr/bin/python2 /usr/bin/supervisord
root        15     1  0 22:03 pts/0    00:00:00 python /usr/bin/supervisor-proc-exit-listener --container-name pmon
root        19     1  0 22:03 pts/0    00:00:00 /usr/sbin/rsyslogd -n -iNONE
root        27     1  0 22:03 pts/0    00:00:00 /usr/bin/python /usr/local/bin/ledd
root        29     1 18 22:03 pts/0    00:00:08 /usr/bin/python /usr/local/bin/xcvrd
root        31     1  1 22:03 pts/0    00:00:00 /usr/bin/python /usr/local/bin/psud
root        40     1  0 22:03 ?        00:00:00 /usr/sbin/sensord -f daemon
root        79    29  0 22:04 pts/0    00:00:00 /usr/bin/python /usr/local/bin/xcvrd
root        83     0  0 22:04 pts/1    00:00:00 bash
root        92    83  0 22:04 pts/1    00:00:00 ps -ef


database0
127.0.0.1:6379[6]> keys *TRAN*
 1) "TRANSCEIVER_DOM_SENSOR|Ethernet12"
 2) "TRANSCEIVER_INFO|Ethernet12"
 3) "TRANSCEIVER_STATUS|Ethernet28"
 4) "TRANSCEIVER_INFO|Ethernet28"
 5) "TRANSCEIVER_DOM_SENSOR|Ethernet20"
 6) "TRANSCEIVER_INFO|Ethernet16"
 7) "TRANSCEIVER_INFO|Ethernet36"
 8) "TRANSCEIVER_STATUS|Ethernet32"
 9) "TRANSCEIVER_STATUS|Ethernet0"
10) "TRANSCEIVER_DOM_SENSOR|Ethernet56"
11) "TRANSCEIVER_DOM_SENSOR|Ethernet16"
12) "TRANSCEIVER_INFO|Ethernet60"
13) "TRANSCEIVER_DOM_SENSOR|Ethernet8"
14) "TRANSCEIVER_STATUS|Ethernet24"
15) "TRANSCEIVER_DOM_SENSOR|Ethernet0"
16) "TRANSCEIVER_DOM_SENSOR|Ethernet36"
17) "TRANSCEIVER_STATUS|Ethernet52"
18) "TRANSCEIVER_STATUS|Ethernet56"
19) "TRANSCEIVER_STATUS|Ethernet36"
20) "TRANSCEIVER_INFO|Ethernet48"
21) "TRANSCEIVER_INFO|Ethernet20"
22) "TRANSCEIVER_INFO|Ethernet8"
23) "TRANSCEIVER_INFO|Ethernet4"
24) "TRANSCEIVER_STATUS|Ethernet12"
25) "TRANSCEIVER_DOM_SENSOR|Ethernet48"
26) "TRANSCEIVER_INFO|Ethernet44"
27) "TRANSCEIVER_DOM_SENSOR|Ethernet32"
28) "TRANSCEIVER_STATUS|Ethernet8"
29) "TRANSCEIVER_INFO|Ethernet52"
30) "TRANSCEIVER_DOM_SENSOR|Ethernet40"
31) "TRANSCEIVER_DOM_SENSOR|Ethernet28"
32) "TRANSCEIVER_INFO|Ethernet56"
33) "TRANSCEIVER_DOM_SENSOR|Ethernet52"
34) "TRANSCEIVER_STATUS|Ethernet48"
35) "TRANSCEIVER_INFO|Ethernet24"
36) "TRANSCEIVER_DOM_SENSOR|Ethernet24"
37) "TRANSCEIVER_INFO|Ethernet32"
38) "TRANSCEIVER_DOM_SENSOR|Ethernet4"
39) "TRANSCEIVER_STATUS|Ethernet16"
40) "TRANSCEIVER_INFO|Ethernet40"
41) "TRANSCEIVER_STATUS|Ethernet20"
42) "TRANSCEIVER_INFO|Ethernet0"
43) "TRANSCEIVER_STATUS|Ethernet4"
44) "TRANSCEIVER_DOM_SENSOR|Ethernet60"
45) "TRANSCEIVER_STATUS|Ethernet60"
46) "TRANSCEIVER_STATUS|Ethernet44"
47) "TRANSCEIVER_STATUS|Ethernet40"
48) "TRANSCEIVER_DOM_SENSOR|Ethernet44"

database1

root@str--acs-1:/# redis-cli -n 6
127.0.0.1:6379[6]> keys *TRANS*
 1) "TRANSCEIVER_STATUS|Ethernet124"
 2) "TRANSCEIVER_DOM_SENSOR|Ethernet68"
 3) "TRANSCEIVER_DOM_SENSOR|Ethernet76"
 4) "TRANSCEIVER_STATUS|Ethernet112"
 5) "TRANSCEIVER_STATUS|Ethernet108"
 6) "TRANSCEIVER_DOM_SENSOR|Ethernet64"
 7) "TRANSCEIVER_DOM_SENSOR|Ethernet108"
 8) "TRANSCEIVER_DOM_SENSOR|Ethernet80"
 9) "TRANSCEIVER_STATUS|Ethernet120"
10) "TRANSCEIVER_STATUS|Ethernet88"
11) "TRANSCEIVER_DOM_SENSOR|Ethernet84"
12) "TRANSCEIVER_INFO|Ethernet72"
13) "TRANSCEIVER_INFO|Ethernet88"
14) "TRANSCEIVER_DOM_SENSOR|Ethernet72"
15) "TRANSCEIVER_STATUS|Ethernet76"
16) "TRANSCEIVER_DOM_SENSOR|Ethernet120"
17) "TRANSCEIVER_STATUS|Ethernet68"
18) "TRANSCEIVER_DOM_SENSOR|Ethernet124"
19) "TRANSCEIVER_INFO|Ethernet120"
20) "TRANSCEIVER_DOM_SENSOR|Ethernet92"
21) "TRANSCEIVER_INFO|Ethernet76"
22) "TRANSCEIVER_DOM_SENSOR|Ethernet88"
23) "TRANSCEIVER_STATUS|Ethernet72"
24) "TRANSCEIVER_INFO|Ethernet80"
25) "TRANSCEIVER_STATUS|Ethernet116"
26) "TRANSCEIVER_INFO|Ethernet108"
27) "TRANSCEIVER_DOM_SENSOR|Ethernet100"
28) "TRANSCEIVER_STATUS|Ethernet104"
29) "TRANSCEIVER_DOM_SENSOR|Ethernet112"
30) "TRANSCEIVER_STATUS|Ethernet64"
31) "TRANSCEIVER_INFO|Ethernet112"
32) "TRANSCEIVER_INFO|Ethernet92"
33) "TRANSCEIVER_INFO|Ethernet104"
34) "TRANSCEIVER_DOM_SENSOR|Ethernet104"
35) "TRANSCEIVER_INFO|Ethernet64"
36) "TRANSCEIVER_STATUS|Ethernet92"
37) "TRANSCEIVER_STATUS|Ethernet84"
38) "TRANSCEIVER_INFO|Ethernet100"
39) "TRANSCEIVER_INFO|Ethernet96"
40) "TRANSCEIVER_DOM_SENSOR|Ethernet96"
41) "TRANSCEIVER_STATUS|Ethernet80"
42) "TRANSCEIVER_INFO|Ethernet124"
43) "TRANSCEIVER_INFO|Ethernet84"
44) "TRANSCEIVER_STATUS|Ethernet96"
45) "TRANSCEIVER_STATUS|Ethernet100"
46) "TRANSCEIVER_INFO|Ethernet68"


admin@str--acs-1:~$ cat /sys/class/leds/port17:green/brightness
255
admin@str--acs-1:~$ sudo ip netns exec asic1 config interface shutdown Ethernet64 
admin@str--acs-1:~$ cat /sys/class/leds/port17:green/brightness
0
admin@str--acs-1:~$ sudo ip netns exec asic1 config interface startup  Ethernet64 
admin@str--acs-1:~$ cat /sys/class/leds/port17:green/brightness
255
```

### CODE REVIEW

File: sonic-ledd/scripts/ledd [hunk 1]
```diff
@@ -17,6 +17,7 @@ try:
     from sonic_daemon_base import daemon_base
     from sonic_daemon_base.daemon_base import Logger
     from sonic_daemon_base.daemon_base import DaemonBase
+    from sonic_daemon_base.daemon_base import multiAsic
```
Comment (jleveque) [inline sonic-ledd/scripts/ledd:20]: This might be for another PR, but if `multiAsic` is a class name, it should be in PascalCase (i.e., `MultiAsic`)

File: sonic-xcvrd/scripts/xcvrd [hunk 2]
```diff
@@ -432,16 +444,23 @@ def del_port_sfp_dom_info_from_db(logical_port_name, int_tbl, dom_tbl):
             sys.exit(NOT_IMPLEMENTED_ERROR)
 
 # recover missing sfp table entries if any
-def recover_missing_sfp_table_entries(sfp_util, int_tbl, status_tbl, stop_event):
+def recover_missing_sfp_table_entries(self, sfp_util, stop_event):
```
Comment (jleveque) [inline sonic-xcvrd/scripts/xcvrd:447]: Since this isn't a method of any class, the first parameter shouldn't be named `self`, *OR* it should be moved to become a class method.

File: sonic-xcvrd/scripts/xcvrd [hunk 3]
```diff
@@ -380,28 +384,36 @@ def post_port_dom_info_to_db(logical_port_name, table, stop_event=threading.Even
...
-
-    appl_db = daemon_base.db_connect("APPL_DB")
-    app_port_tbl = swsscommon.ProducerStateTable(appl_db,
-                                                 swsscommon.APP_PORT_TABLE_NAME)
+    transceiver_dict, state_db, appl_db, int_tbl, dom_tbl, app_port_tbl = {}, {}, {}, {}, {}, {}
+
+    # Get the namespaces in the platform
+    namespaces = multi_asic.get_namespaces()
+    for namespace in namespaces:
+        asic_id = multi_asic.get_asic_id_from_namespace(namespace)
```
Comment (keboliu) [inline sonic-xcvrd/scripts/xcvrd:392]: could you point me the design for multi basic?  From code here my understanding is that each ASIC "own" a set of ports/transceiver, I would like to understand more.

File: sonic-ledd/scripts/ledd (hunk 1 above)
Comment (judyjoseph) [reply sonic-ledd/scripts/ledd]: Changed the name to MultiAsic and updated the other PR as well.

File: sonic-xcvrd/scripts/xcvrd (hunk 2 above)
Comment (judyjoseph) [reply sonic-xcvrd/scripts/xcvrd]: Thanks Joe, I have reverted back to use int_tbl, status_tbl itself. The int_tbl, status_tbl are now list datatype and indexed by the asic_id. In case of single ASIC patform the asic_id is 0 , ( returned by the get_asic_id_from_namespace() API in MultiAsic class in daemon_base.py )

File: sonic-xcvrd/scripts/xcvrd (hunk 3 above)
Comment (judyjoseph) [reply sonic-xcvrd/scripts/xcvrd:392]: > could you point me the design for multi basic? From code here my understanding is that each ASIC "own" a set of ports/transceiver, I would like to understand more.

The design is explained in brief in the multi-ASIC HLD, but I have an internal document on the design approaches, shall share it with you.

File: sonic-ledd/scripts/ledd [hunk 4]
```diff
@@ -75,15 +79,20 @@ class DaemonLedd(DaemonBase):
             logger.log_error("Failed to load ledutil: %s" % (str(e)), True)
             sys.exit(LEDUTIL_LOAD_ERROR)
 
-        # Open a handle to the Application database
-        appl_db = daemon_base.db_connect("APPL_DB")
+        # Get the namespaces in the platform
+        namespaces = multi_asic.get_namespaces()
```
Comment (sujinmkang) [inline sonic-ledd/scripts/ledd:83]: if it's not a multi-asic platform, does this return only one namespace?

File: sonic-ledd/scripts/ledd [hunk 5]
```diff
@@ -96,17 +105,19 @@ class DaemonLedd(DaemonBase):
...
                 continue
 
-            (key, op, fvp) = sst.pop()
-
-            # TODO: Once these flag entries have been removed from the DB,
-            # we can remove this check
-            if key in ["PortConfigDone", "PortInitDone"]:
-                continue
+            for namespace in namespaces:
+                (key, op, fvp) = sst[namespace].pop()
```
Comment (abdosi) [inline sonic-ledd/scripts/ledd:109]: i think we can do this c.getdbId().namespace

File: sonic-xcvrd/scripts/xcvrd (hunk 3 above)
Comment (judyjoseph) [reply sonic-xcvrd/scripts/xcvrd:392]: I have moved the design into a separate design doc, https://github.com/judyjoseph/SONiC/blob/multi_asic/doc/pmon/pmon_multiasic_design.md. Let me know if you are not able to access.

File: sonic-ledd/scripts/ledd (hunk 4 above)
Comment (judyjoseph) [reply sonic-ledd/scripts/ledd:83]: yes, if it is not a multi-asic platform it returns back "" EMPTY_NAMESPACE which signifies linux host. 
Please take a look at the related PR https://github.com/Azure/sonic-buildimage/pull/4932

File: sonic-xcvrd/scripts/xcvrd (hunk 3 above)
Comment (judyjoseph) [reply sonic-xcvrd/scripts/xcvrd:392]: @keboliu, were you able to go through the approach doc ? (https://github.com/Azure/SONiC/pull/655) 
Could you share your comments.

File: sonic-xcvrd/scripts/xcvrd [hunk 6]
```diff
@@ -95,10 +96,35 @@ platform_chassis = None
...
 
+# The empty namespace refers to linux host namespace.
+EMPTY_NAMESPACE = ''
+
 #
 # Helper functions =============================================================
 #
 
+# Get namespaces, for single ASIC platform namespace is EMPTY_NAMESPACE
+def get_front_end_namesapces():
```
Comment (jleveque) [inline sonic-xcvrd/scripts/xcvrd:107]: Typo: s/get_front_end_namesapces/get_front_end_namespaces/

Also, this function seems like a good candidate function for sonic-py-common. No?

File: sonic-xcvrd/scripts/xcvrd [hunk 7]
```diff
@@ -95,10 +96,35 @@ platform_chassis = None
...
+    namespaces = [EMPTY_NAMESPACE]
+    if multi_asic.is_multi_asic():
+        ns_list = multi_asic.get_all_namespaces()
+        namespaces = ns_list['front_ns']
+
+    return namespaces
+
+# Get asic index from the namespace name
+# With single ASIC platform, namespace is EMPTY_NAMESPACE, return index 0
+def get_asic_id_from_namespace(namespace):
```
Comment (jleveque) [inline sonic-xcvrd/scripts/xcvrd:119]: This function seems like a good candidate function for sonic-py-common. No?

File: sonic-xcvrd/scripts/xcvrd (hunk 6 above)
Comment (judyjoseph) [reply sonic-xcvrd/scripts/xcvrd:107]: These API's are actually wrapper API's which will get the ( for eg: front end namespaces ) irrespective of whether it is a single asic or multi-asic platform. Yes I can definitely move them to multi_asic.py. 

Currently multi_asic.py .. have the API's used by only platforms with >1 ASIC ) -- May we could have some comments in the same file and start adding API's which could be used by both single/multi-asic platforms.

Comment (jleveque) [reply sonic-xcvrd/scripts/xcvrd:107]: I think that makes sense. Do you intend to commit this as-is, then move the functions and remove them from here later, or open a PR to add them to sonic-py-common first, then remove them from this PR?

Comment (judyjoseph) [reply sonic-xcvrd/scripts/xcvrd:107]: let me open a new PR add these to sonic-py-common first. Later update this PR ... that should be clean.

Comment (judyjoseph) [reply sonic-xcvrd/scripts/xcvrd:107]: Raised this PR, to add these API's ... https://github.com/Azure/sonic-buildimage/pull/5224

File: sonic-xcvrd/scripts/xcvrd [hunk 8]
```diff
@@ -719,28 +738,40 @@ def detect_port_in_error_status(logical_port_name, status_tbl):
 # Init TRANSCEIVER_STATUS table
 def init_port_sfp_status_tbl(stop_event=threading.Event()):
     # Connect to STATE_DB and create transceiver status table
-    state_db = daemon_base.db_connect("STATE_DB")
-    status_tbl = swsscommon.Table(state_db, TRANSCEIVER_STATUS_TABLE)
+    state_db, status_tbl = {},{}
```
Comment (jleveque) [inline sonic-xcvrd/scripts/xcvrd:741]: Add space after comma: `{}, {}`

Comment (judyjoseph) [reply sonic-xcvrd/scripts/xcvrd]: Updated Joe.

**Comment by abdosi** [comment]
@judyjoseph create PR for 201911

**Comment by judyjoseph** [comment]
> @judyjoseph create PR for 201911

Raised this PR  https://github.com/Azure/sonic-platform-daemons/pull/89
